gcc/config/riscv/bitmanip.md

```diff
@@ -83,6 +83,74 @@
  
 ;;; ??? pack
 
+(define_insn "riscv_bitmanip_pack<mode>"
+  [(set (match_operand:X 0 "register_operand" "=r")
+	(unspec
+	    [(match_operand:X 1 "register_operand" "r")
+		 (match_operand:X 2 "register_operand" "r")]
+	    UNSPEC_PACK))]
+  "TARGET_ZBP"
+  "pack\t%0,%1,%2"
+  [(set_attr "type" "bitmanip")
+   (set_attr "length" "4")])
+(define_insn "riscv_bitmanip_packu<mode>"
+  [(set (match_operand:X 0 "register_operand" "=r")
+	(unspec
+	    [(match_operand:X 1 "register_operand" "r")
+		 (match_operand:X 2 "register_operand" "r")]
+	    UNSPEC_PACKU))]
+  "TARGET_ZBP"
+  "packu\t%0,%1,%2"
+  [(set_attr "type" "bitmanip")
+   (set_attr "length" "4")])
+
+;; By the time we reach this, gcc has changed
+;; `(a&0ff)<<8` into `(a<<8)&0xFF00`, despite the fact
+;; it's worse on RISC-V (need to set up the constant
+;; for the shift vs. andi/slli)
+;; Keep an easily recognizable subtree
+(define_insn "*extractB0toB1"
+[(set (match_operand:SI 0 "register_operand" "=r")
+      (and:SI (ashift:SI (match_operand:SI 1 "register_operand" "r")
+	                     (const_int 8))
+              (const_int 65280)))]
+"!TARGET_64BIT"
+"andi\\t%0,%1,0xff\;slli\\t%0,%0,0x8"
+[(set_attr "length" "8")]
+)
+(define_insn "*extractB0toB2"
```

Comment:
I would prefer `define_split` or `define_insn_and_split` for `extractB0toB1` and `extractB0toB2`, although I think this should fixed in front-end or middle-end, but this fix is much simpler than modify front-end/middle-end,

Reply:
done